wgengine/packet/packet.go

```diff
@@ -37,11 +40,53 @@ func (p IPProto) String() string {
 		return "UDP"
 	case TCP:
 		return "TCP"
+	case IPv6:
+		return "IPv6"
 	default:
 		return "Junk"
 	}
 }
 
+// These are the minimal sizes of the respective headers.
+// It is therefore not appropriate to use them during decoding.
+// They are used in packet assembly, where we always have minimal headers.
+const (
+	IPHeaderSize   = 20
+	ICMPHeaderSize = 4
+	TCPHeaderSize  = 20
+	UDPHeaderSize  = 8
+)
+
+// These are the offsets at which each protocol's payload starts.
+const (
+	ICMPDataOffset = IPHeaderSize + ICMPHeaderSize
```

Comment:
Whose are these exported for?

Reply:
This is so that callers know where to write the payload. The payload has to be written before headers because it's checksummed, so things typically look like
```go
dns.pack(out[transportOffset+DNSDataOffset:])
WriteDNSHeader(out[transportOffset:])
tun.Injectinbound(out, transportOffset)
```

Reply:
Ah. Maybe we should have some packet builder type that we maintain in a sync.Pool for garbage reasons.

Then that packet builder owns the underlying packet buffer and we can Write/WriteByte into it at the right place (regardless of IP header size) and it'll checksum in the right place when we're done.

Reply:
I'm implementing this right now and it looks fine, but then these builders have to know about Wireguard's outer transport headers, because `tun.Device` (and so also `tstun.TUN`) requires at least that much leading space in a buffer passed to Read/Write.

Reply:
Maybe headers can return their `Size`? Then it can vary according to, say, IPv4/6, if I understand the source of your concern correctly.